File: termin/cpp/termin/render_target_config.cpp
// render_target_config.cpp - RenderTargetConfig implementation
#include "render_target_config.hpp"
#include <tcbase/tgfx_intern_string.h>

namespace termin {

tc_render_target_config RenderTargetConfig::to_c() const {
    tc_render_target_config c;
    tc_render_target_config_init(&c);

    c.name = name.empty() ? nullptr : tgfx_intern_string(name.c_str());
    c.camera_uuid = camera_uuid.empty() ? nullptr : tgfx_intern_string(camera_uuid.c_str());
    c.width = width;
    c.height = height;
    c.pipeline_uuid = pipeline_uuid.empty() ? nullptr : tgfx_intern_string(pipeline_uuid.c_str());
    c.pipeline_name = pipeline_name.empty() ? nullptr : tgfx_intern_string(pipeline_name.c_str());
    c.layer_mask = layer_mask;
    c.enabled = enabled;

    return c;
}

RenderTargetConfig RenderTargetConfig::from_c(const tc_render_target_config* c) {
    RenderTargetConfig cfg;
    if (!c) return cfg;

    cfg.name = c->name ? c->name : "";
    cfg.camera_uuid = c->camera_uuid ? c->camera_uuid : "";
    cfg.width = c->width;
    cfg.height = c->height;
    cfg.pipeline_uuid = c->pipeline_uuid ? c->pipeline_uuid : "";
    cfg.pipeline_name = c->pipeline_name ? c->pipeline_name : "";
    cfg.layer_mask = c->layer_mask;
    cfg.enabled = c->enabled;

    return cfg;
}

} // namespace termin

File: termin-app/termin/visualization/core/render_target_config.py
"""RenderTargetConfig - configuration for scene render targets."""

from __future__ import annotations

from termin.entity._entity_native import RenderTargetConfig

__all__ = ["RenderTargetConfig", "serialize_render_target_config", "deserialize_render_target_config"]


def serialize_render_target_config(config: RenderTargetConfig) -> dict:
    """Serialize RenderTargetConfig to dict."""
    result = {
        "name": config.name,
        "camera_uuid": config.camera_uuid,
    }
    if config.dynamic_resolution:
        result["dynamic_resolution"] = True
    else:
        result["width"] = config.width
        result["height"] = config.height
    if config.color_format:
        result["color_format"] = config.color_format
    if config.depth_format:
        result["depth_format"] = config.depth_format
    if config.pipeline_uuid:
        result["pipeline_uuid"] = config.pipeline_uuid
    if config.pipeline_name:
        result["pipeline_name"] = config.pipeline_name
    if config.layer_mask != 0xFFFFFFFFFFFFFFFF:
        result["layer_mask"] = hex(config.layer_mask)
    if not config.enabled:
        result["enabled"] = config.enabled
    if config.pipeline_params:
        result["pipeline_params"] = dict(config.pipeline_params)
    return result


def deserialize_render_target_config(data: dict) -> RenderTargetConfig:
    """Deserialize RenderTargetConfig from dict."""
    config = RenderTargetConfig()
    config.name = data.get("name", "")
    config.camera_uuid = data.get("camera_uuid", "")
    config.width = data.get("width", 512)
    config.height = data.get("height", 512)
    config.dynamic_resolution = data.get("dynamic_resolution", False)
    config.color_format = data.get("color_format", "rgba16f")
    config.depth_format = data.get("depth_format", "depth32f")
    config.pipeline_uuid = data.get("pipeline_uuid", "")
    config.pipeline_name = data.get("pipeline_name", "")

    layer_mask_raw = data.get("layer_mask", 0xFFFFFFFFFFFFFFFF)
    if isinstance(layer_mask_raw, str):
        config.layer_mask = int(layer_mask_raw, 16) if layer_mask_raw.startswith("0x") else int(layer_mask_raw)
    else:
        config.layer_mask = int(layer_mask_raw)

    config.enabled = data.get("enabled", True)
    pipeline_params = data.get("pipeline_params", {})
    if isinstance(pipeline_params, dict):
        config.pipeline_params = {
            str(slot): str(value)
            for slot, value in pipeline_params.items()
            if slot and value
        }
    return config
